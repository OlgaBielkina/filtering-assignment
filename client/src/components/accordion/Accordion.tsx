import React, { useState } from 'react';

interface Props {
    children: React.ReactNode;
}

export interface IAccordionContext {
    expanded: boolean;
    handleExpanded?: () => void;
}

export const AccordionContext = React.createContext<IAccordionContext>({
    expanded: false,
    handleExpanded: undefined,
});

export const Accordion: React.FC<Props> = ({ children }) => {
    const [expanded, setExpanded] = useState(false);
    const handleExpanded = () => setExpanded(!expanded);
    const data = { expanded, handleExpanded };

    return <AccordionContext.Provider value={data}>{children}</AccordionContext.Provider>;
};
