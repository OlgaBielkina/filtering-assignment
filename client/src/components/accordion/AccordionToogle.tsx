import React, { useContext } from 'react';
import classnames from 'classnames';
import { AccordionContext } from './Accordion';
import styles from './AccordionToogle.module.scss';

export interface Props {
    children: React.ReactNode;
    collapsible: boolean;
}

export const AccordionToogle: React.FC<Props> = ({ children, collapsible }) => {
    const context = useContext(AccordionContext);
    const handler = (e: React.MouseEvent<HTMLDivElement>) => {
        e.stopPropagation();
        if (collapsible && context.handleExpanded) {
            context.handleExpanded();
        }
    };

    return (
        <div className={styles.root} onClick={handler}>
            {children}
            {collapsible && (
                <span className={classnames('material-icons', styles.icon)}>
                    {context.expanded ? 'expand_less' : 'expand_more'}
                </span>
            )}
        </div>
    );
};
