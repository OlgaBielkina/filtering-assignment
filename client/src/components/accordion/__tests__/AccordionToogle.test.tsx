import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { AccordionToogle, Props } from '../AccordionToogle';
import { AccordionContext, IAccordionContext } from '../Accordion';

describe('AccordionToogle', () => {
    const props: Omit<Props, 'children'> = {
        collapsible: true,
    };
    const providerData: IAccordionContext = {
        expanded: false,
        handleExpanded: jest.fn(),
    };

    it('should render collapsed toggle', () => {
        const { asFragment, getByText } = render(
            <AccordionContext.Provider value={providerData}>
                <AccordionToogle {...props}>test text</AccordionToogle>
            </AccordionContext.Provider>,
        );

        getByText('expand_more');
        expect(asFragment()).toMatchSnapshot();
    });

    it('should render expanded toggle', () => {
        const { asFragment, getByText } = render(
            <AccordionContext.Provider value={{ ...providerData, expanded: true }}>
                <AccordionToogle {...props}>test text</AccordionToogle>
            </AccordionContext.Provider>,
        );

        getByText('expand_less');
        expect(asFragment()).toMatchSnapshot();
    });

    it('should not render collapsible toggle', () => {
        const { asFragment, queryByText } = render(
            <AccordionContext.Provider value={providerData}>
                <AccordionToogle {...{ ...props, collapsible: false }}>test text</AccordionToogle>
            </AccordionContext.Provider>,
        );

        expect(queryByText('expand_less')).not.toBeInTheDocument();
        expect(queryByText('expand_more')).not.toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    it('should chnage icon when user press toggle', () => {
        const { getByText } = render(
            <AccordionContext.Provider value={providerData}>
                <AccordionToogle {...props}>test text</AccordionToogle>
            </AccordionContext.Provider>,
        );

        fireEvent.click(getByText('expand_more'));

        expect(providerData.handleExpanded).toBeCalledTimes(1);
    });
});
