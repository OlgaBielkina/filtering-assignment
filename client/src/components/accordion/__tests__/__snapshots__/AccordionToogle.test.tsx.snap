// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionToogle should not render collapsible toggle 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    test text
  </div>
</DocumentFragment>
`;

exports[`AccordionToogle should render collapsed toggle 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    test text
    <span
      class="material-icons icon"
    >
      expand_more
    </span>
  </div>
</DocumentFragment>
`;

exports[`AccordionToogle should render expanded toggle 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    test text
    <span
      class="material-icons icon"
    >
      expand_less
    </span>
  </div>
</DocumentFragment>
`;
